torch_geometric/explain/metrics.py

```diff
@@ -0,0 +1,125 @@
+        metrics."""
+
+    @abstractmethod
+    def compute_metric(self):
+        r"""Compute the metric over all inputs."""
+
+    @abstractmethod
+    def aggregate(self):
+        r"""Aggregate metrics over all inputs"""
+
+
+class Fidelity(ExplanationMetric):
+    r"""Fidelity+/- Explanation Metric as
+    defined in https://arxiv.org/abs/2206.09677"""
+    def __init__(self) -> None:
+        super().__init__()
+
+
+def fidelity(explainer: Explainer, explanation: Explanation,
+             target: Optional[Tensor] = None,
+             index: Optional[Union[int, Tensor]] = None,
+             output_type: str = 'raw', **kwargs):
+    r"""Evaluate the fidelity of Explainer and given
+    explanation produced by explainer
+
+    Args:
+        explainer :obj:`~torch_geometric.explain.Explainer`
+            The explainer to evaluate
+        explanation :obj:`~torch_teometric.explain.Explanation`
+            The explanation to evaluate
+        target (Tensor, optional): The target prediction, if not provided it
+            is inferred from obj:`explainer`, defaults to obj:`None`
+        index (Union[int, Tensor]): The explanation target index, for node-
+            and edge- level task it signifies the nodes/edges explained
+            respectively, for graph-level tasks it is assumed to be None,
+            defaults to obj:`None`
+    """
+    metric_dict = {}
+
+    task_level = explainer.model_config.task_level
+
+    if index != explanation.get('index'):
+        raise ValueError(f'Index ({index}) does not match '
+                         f'explanation.index ({explanation.index}).')
+
+    # get input graphs
+    explanation_graph = explanation.get_explanation_subgraph()  # for fid-
+    complement_graph = explanation.get_complement_subgraph()  # for fid+
+
+    # get target
+    target = explainer.get_target(x=explanation.x,
+                                  edge_index=explanation.edge_index,
+                                  **kwargs)  # using full explanation
+
+    # get predictions
+    explanation_prediction = explainer.get_prediction(
+        x=explanation_graph.x, edge_index=explanation_graph.edge_index,
+        **kwargs)
+    complement_prediction = explainer.get_prediction(
+        x=complement_graph.x, edge_index=complement_graph.edge_index, **kwargs)
+
+    # fix logprob to prob
+    if output_type == 'prob' and explainer.model.return_type == 'log_probs':
+        target = torch.exp(target)
+        explanation_prediction = torch.exp(explanation_prediction)
+        complement_prediction = torch.exp(complement_prediction)
+
+    # based on task level
+    if task_level == 'graph':
+        if index is not None:
+            ValueError(
+                f'Index for graph level task should be None, got (f{index})')
+        # evaluate whole entry
+        pass
+    elif task_level == 'edge':
+        # get edge prediction
+        pass  # TODO (blaz)
```

Comment:
Node level and graph level tasks make sense to me. For nodes we will have a clearly indexed output (each row of the prediction corresponding to nodes) and for full graphs we only have one prediction anyway. This makes it clear on what to compare to get fidelities (both in classification and regression tasks). 

But I am not certain for edge level tasks. Let say for example when obtaining the explanation for an edge level task with the explainer class, we use:
```
explanation = explainer(
        x,
        edge_index,
        target=target,
        index=index,
        edge_label_index=edge_label_index,
    )
```
What exactly does the `index` refer to, are we explaining the prediciton(s) for edge(s) with indices in `index` which correspond to edges in `edge_index`? Or something else? If so, how can we easily compare the predictions, as the `edge_index` changes for induced and complement graphs for an explanation?

Reply:
Maybe the tests can help ? https://github.com/pyg-team/pytorch_geometric/blob/b2f9d0521f687e8f5341b998f52227540bab99fe/test/explain/algorithm/test_gnn_explainer.py#L84

I haven't taken a look at the edge level explainers in details, but it seems that for explaining the edge between node `0` and  `1` you would have `edge_label_index = torch.tensor([[0],1])`.

Reply:
@BlazStojanovic I think we should leave this as `NotImplemented` for now. Because I think `task_level=edge` needs some more work
1. We don't have an example to demonstrate `task_level=edge` on some real world data. We could have and example that explains [this](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/rgcn_link_pred.py#L20) edge prediction task. 
2. GNNExplainer needs to be updated to support `edge` level tasks, one line that needs to be updated is https://github.com/pyg-team/pytorch_geometric/blob/f622fee0d1ab71d1b2f97f9cc31d7990b45d99d9/torch_geometric/explain/algorithm/gnn_explainer.py#L116.